packages/builtin/iterator.pony

```diff
@@ -1,3 +1,85 @@
 interface Iterator[A]
+  """
+
+  Iterators generate a series of values, one value at a time on each call to `next()`.
+
+  An Iterator is considered exhausted, once its `has_next()` method returns `false`.
+  Thus every call to `next()` should be preceeded with a call to `has_next()` to
+  check for exhaustiveness.
+
+  ## Usage
+
+  Given the rules for using Iterators mentioned above, basic usage
+  of an iterator looks like this:
+
+  ```pony
+  while iterator.has_next() do
+    let elem = iterator.next()?
+    // do something with elem
+  end
+  ```
+
+  The `For`-loop provides a more concise way of iteration:
+
+  ```pony
+  for elem in iterator do
+    // do something with elem
+  end
+  ```
+
+  Iteration using `While` is more flexible as it allows to continue iterating if a call to `next()` errors.
+  The `For`-loop does not allow this.
+
+  ## Implementing Iterators
+
+  Iterator implementations need to adhere to the following rules to be considered well-behaved:
+
+  * If the Iterator is exhausted, `has_next()` needs to return `false`.
+  * Once `has_next()` returned `false` it is not allowed to switch back to `true`
+    (Unless the Iterator supports rewinding)
+  * `has_next()` does not change its returned value if `next()` has not been called.
+    That means, that between two calls to `next()` any number of calls to `has_next()`
+    need to return the same value. (Unless the Iterator supports rewinding)
+  * A call to `next()` erroring does not necessarily denote exhaustiveness.
+
+  ### Example
+
+  ```pony
+  // Generates values from `from` to 0
+  class ref Countdown is Iterator[USize]
+    var _cur: USize
+    var _has_next: Bool = true
+
+    new ref create(from: USize) =>
+      _cur = from
+
+    fun ref has_next(): Bool =>
+      _has_next
+
+    fun ref next(): USize =>
+      let elem = _cur = _cur - 1
+      if elem == 0 then
+        _has_next = false
+      end
+      elem
+  ```
+
+  """
+
   fun ref has_next(): Bool
+    """
+    Returns `true` if this Iterator is not yet exhausted.
+    That means a subsequent call to `next()` will return a value.
```

Comment:
As stated in the other docstrings that you are adding, subsequent calls to next may raise an error, even if `has_next` returns true.

Reply:
true. I hope the fixed formulation works. I am not that good at writing docs.